alchemiscale/storage/statestore.py

```diff
@@ -602,8 +603,161 @@ def get_scoped_key(self, obj: GufeTokenizable, scope: Scope):
         return res[0]
 
     def get_gufe(self, scoped_key: ScopedKey):
-        node, subgraph = self._get_node(scoped_key=scoped_key, return_subgraph=True)
-        return self._subgraph_to_gufe([node], subgraph)[node]
+        return self.get_keyed_chain(scoped_key).to_gufe()
+
+    def get_keyed_chain(self, scoped_key: ScopedKey) -> KeyedChain:
+        """Retrieve the ``KeyedChain`` form of a ``GufeTokenizable`` from the database.
+
+        Parameters
+        ----------
+        scoped_key
+            The ``ScopedKey`` of the ``GufeTokenizable`` to retrieve.
+
+        Returns
+        -------
+        ``KeyedChain``
+            The ``KeyedChain`` form of the tokenizable.
+        """
+
+        # find the root node and all nodes that are connected by any number of
+        # "DEPENDS_ON" relationships. Discard the path object, just collecting
+        # the child nodes into a single set of nodes. Unwind all nodes and get
+        # their dependencies, both the data embedded in the relationship,
+        # and their keys.
+        query = """
+        MATCH (root:GufeTokenizable {`_scoped_key`: $scoped_key })
+        OPTIONAL MATCH (root)-[:DEPENDS_ON*]->(dep)
+        WITH root, COLLECT(DISTINCT dep) AS deps
```

Comment:
This makes much more sense to me now, since it means below this we are now effectively looping through each relevant `GufeTokenizable` (`root` and all its downward dependencies) only once.